lib/nmatrix/shortcuts.rb

```diff
@@ -465,7 +465,137 @@ def random(shape, opts={})
+    #    NMatrix.magic(6) # => [  [31,  9,  2, 22, 27, 20]
+    #                             [ 3, 32,  7, 21, 23, 25]
+    #                             [35,  1,  6, 26, 19, 24]
+    #                             [ 4, 36, 29, 13, 18, 11]
+    #                             [30,  5, 34, 12, 14, 16]
+    #                             [ 8, 28, 33, 17, 10, 15] ]
+    #
+    def magic(shape)
+      if shape==2
+        raise(ArgumentError, "shape of two is not allowed")
+      end
+      nm= NMatrix.new([shape,shape],0)
+      if shape%2!=0
+        odd_magic(nm,shape)
+      elsif shape%4 == 0
+        doubly_even_magic(nm,shape)
+      else   
+        singly_even_magic(nm,shape)
+      end
+      nm
+    end
+    
+    def odd_magic(nm,shape)
+      row= shape-1
+      col= shape/2       
+      nm[row,col]= 1
+      (2..shape*shape).each do |i|
+        if nm[(row+1)%shape,(col+1)%shape] == 0
+          row= (row+1)%shape
+          col= (col+1)%shape
+        else
+          row= (row-1+shape)%shape
+        end
+          nm[row,col]= i
+      end
+    end
+    
+    def doubly_even_magic(nm,shape)
+      mini_square_num = shape/4
+      cnt = 1     
+      inv_cnt = shape*shape
+      shape.times do |i|
+        shape.times do |j|
+          if j>=mini_square_num && j<shape-mini_square_num
+		    if i>=mini_square_num && i<shape-mini_square_num
+      	      nm[i,j] = cnt
+		    else 
+		      nm[i,j] = inv_cnt
+		    end
+		  elsif i<mini_square_num || i>=shape-mini_square_num
+		    nm[i,j]=cnt
+		  else
+            nm[i,j] = inv_cnt
+          end
+        cnt+=1
+        inv_cnt-=1  
+        end
+      end
+    end
+    
+    def singly_even_magic(nm,shape)
+      half_shape = shape/2
+	  k = (shape-2)/4
+	  swap_col= NMatrix.new([shape], dtype: :int32)
+      index=0 
+      mini_magic= NMatrix.new([half_shape,half_shape],0)
+      odd_magic(mini_magic,half_shape)
+      half_shape.times do |i|
+        half_shape.times do |j|
+          nm[i,j] = mini_magic[i,j]  	
+      	  nm[i+half_shape,j+half_shape] = mini_magic[i,j]+half_shape*half_shape  
+          nm[i,j+half_shape] = mini_magic[i,j]+2*half_shape*half_shape      
+          nm[i+half_shape,j] = mini_magic[i,j]+3*half_shape*half_shape       
+        end  
+      end
+  
+      (1..k).each do |i|
+        swap_col[index] = i
+        index+=1
+      end
+      
+      (shape-k+2..shape).each do |i|
+        swap_col[index] = i
+        index+=1
+      end 
+      
+      (1..half_shape).each do |i|
+        (1..index).each do |j|
+          temp=nm[i-1,swap_col[j-1]-1]
+          nm[i-1,swap_col[j-1]-1]=nm[i+half_shape-1,swap_col[j-1]-1]
+          nm[i+half_shape-1,swap_col[j-1]-1]=temp
+        end
+      end
+
+      temp=nm[k,0] 
+      nm[k,0]=nm[k+half_shape,0] 
+      nm[k+half_shape,0]=temp
 
+      temp=nm[k+half_shape,k]
+      nm[k+half_shape,k]=nm[k,k] 
+      nm[k,k]=temp
+    end   
+    private :odd_magic, :doubly_even_magic, :singly_even_magic
```

Comment:
So you can create a module `MagicHelpers` inside this same file and add the functions listed as private to the module as class methods.

So:
``` ruby
class NMatrix
  module MagicHelpers
    class << self
      # methods
    end
  end
end

def self.magic n
  # call as MagicHelpers.method
end
```

@mohawkjohn this is what you had told me to do when I was writing a function for solving tridiagonal equations. That PR got old so I closed it: https://github.com/SciRuby/nmatrix/pull/301/files